File: src/main/scala/chapter5/scala/KeyedTransformations.scala
package chapter5.scala

import org.apache.flink.api.common.eventtime.{SerializableTimestampAssigner, WatermarkStrategy}
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment
import util.scala.{SensorReading, SensorSource}

import java.time.Duration

object KeyedTransformations {

    def main(args: Array[String]): Unit = {

        // set up the streaming execution environment
        val env = StreamExecutionEnvironment.getExecutionEnvironment()

        val strategy: WatermarkStrategy[SensorReading] = WatermarkStrategy
          .forBoundedOutOfOrderness(Duration.ofSeconds(5L))
          .withTimestampAssigner(new SerializableTimestampAssigner[SensorReading] {
            override def extractTimestamp(t: SensorReading, l: Long): Long = t.timestamp
          })

        // ingest sensor stream
        val readings = env
          .addSource(new SensorSource())
          .assignTimestampsAndWatermarks(strategy)

        // group sensor readings by sensor id
        val keyed = readings.keyBy((in: SensorReading) => in.id)

        // a rolling reduce that computes the highest temperature of each sensor and the corresponding timestamp

        val maxTempPerSensor = keyed
          .reduce((r1, r2) => if (r1.temperature > r2.temperature) r1 else r2)

        maxTempPerSensor.print()

        // execute application
        env.execute("Keyed Transformations Example")
    }
}

File: src/main/java/chapter5/KeyedTransformations.java
package chapter5;

import org.apache.flink.api.common.eventtime.WatermarkStrategy;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.datastream.KeyedStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import util.SensorReading;
import util.SensorSource;

import java.time.Duration;

public class KeyedTransformations {

    public static void main(String[] args) throws Exception {

        // set up the streaming execution environment
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();

        WatermarkStrategy<SensorReading> strategy = WatermarkStrategy
                .<SensorReading>forBoundedOutOfOrderness(Duration.ofSeconds(5L))
                .withTimestampAssigner((sr, ts) -> sr.getTimestamp());

        // ingest sensor stream
        DataStream<SensorReading> readings = env
                // SensorSource generates random temperature readings
                .addSource(new SensorSource())
                // assign timestamps and watermarks which are required for event time
                .assignTimestampsAndWatermarks(strategy);

        // group sensor readings by sensor id
        KeyedStream<SensorReading, String> keyed = readings
                .keyBy(SensorReading::getId);

        // a rolling reduce that computes the highest temperature of each sensor and
        // the corresponding timestamp
        DataStream<SensorReading> maxTempPerSensor = keyed
                .reduce((r1, r2) -> {
                    if (r1.getTemperature() > r2.getTemperature()) {
                        return r1;
                    } else {
                        return r2;
                    }
                });

        maxTempPerSensor.print();

        // execute application
        env.execute("Keyed Transformations Example");
    }
}